Implement TargetOptions class

ffoulkes opened the pull request:
Replace the compile-time conditionals in ConfigMonitoringService and P4Service with run-time tests on a parameter object (TargetOptions) that is specified when the service is instantiated.

Use TargetOptions to control whether:

1. ConfigMonitoringService requires that the chassis config file not be a symlink. (enabled for security)
2. ConfigMonitoringService pushes an updated chassis configuration. (disabled for DPDK)
3. P4Service allows a configured pipeline to be overwritten. (disabled for DPDK)

Update DpdkMain to pass a TargetOptions object to DpdkHal, which in turn passes it to ConfigMonitoringService and P4Service. Modify constructors and factory methods to accept the TargetOption object while maintaining backward compatibility with existing code.

The secureChassisConfig option is currently disabled, to support unit testing. Further work is needed to allow this option to be enabled by default for TDI targets and disabled by default for non-TDI targets and unit tests.

ffoulkes committed:
Implement TargetOptions class

Replace the compile-time conditionals in ConfigMonitoringService and
P4Service with run-time tests on a parameter object (TargetOptions)
that is specified when the service is instantiated.

Use TargetOptions to control whether:

1) ConfigMonitoringService requires that the chassis config file
   not be a symlink. (enabled for security)
2) ConfigMonitoringService pushes an updated chassis configuration.
   (disabled for DPDK)
3) P4Service allows a configured pipeline to be overwritten.
   (disabled for DPDK)

Update DpdkMain to pass a TargetOptions object to DpdkHal, which in
turn passes it to ConfigMonitoringService and P4Service. Modify
constructors and factory methods to accept the TargetOption object
while maintaining backward compatibility with existing code.

The secureChassisConfig option is currently disabled, to support
unit testing. Further work is needed to allow this option to be
enabled by default for TDI targets and disabled by default for
non-TDI targets and unit tests.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes commented:
This eliminates the `#ifdef DPDK_TARGET` conditionals in `config_monitoring_service.cc` and `p4_service.cc` and the need to supply those conditionals at compile time.

If we can eliminate the need to specify `--define target=<target>` Bazel command lines,
1.  The commands will be easier to issue,
2. The BUILD files will be easier to write and maintain, and
3. Bazel won't have to recompile files just we changed a command-line parameter.

The latter will help speed up builds.

It should also improve our chances of being able to upstream our changes to the common files.

ffoulkes commented:
I also ran [buildifier](https://github.com/bazelbuild/buildtools/blob/master/buildifier/README.md) on the common BUILD file. It applies standard formatting conventions to Bazel BUILD and .bzl files. We will eventually want to do this for all the Bazel files we create or modify.

5abeel commented on stratum/hal/lib/common/target_options.h:
```diff
@@ -0,0 +1,33 @@
+
+  // Whether ConfigMonitoringService should require that the config file
+  // be a regular file (not a symlink).
+  bool secureChassisConfig;
```
Why/when would we allow a symlink ChassisConfig file? What is the use case for this?

ffoulkes replied:
> Why/when would we allow a symlink ChassisConfig file? What is the use case for this?

Two use cases:

1. One of the unit tests uses a symlink. The check breaks the unit test.
2. We need to disable the symlink check for non-TDI targets, because it changes the behavior of existing code. The check must be conditionalized somehow, and I'm trying to do it without #ifdefs, which are a pain in Bazel.

This is what happens when we edit common files.

ffoulkes committed:
Rebase target_options branch

- Rebase target_options branch on split_arch branch.

- Add target_options source files to stratum cmake project.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Remove DPDK_TARGET and ES2K_TARGET

- Remove definitions of DPDK_TARGET and ES2K_TARGET from the
  Bazel build, leaving just TOFINO_TARGET.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
TargetOptions support

- Finish implementing TargetOptions support for DPDK and ES2K.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes commented:
Completed the implementation for both DPDK and ES2K. Removed definitions of DPDK_TARGET and ES2K_TARGET from the Bazel build.

ffoulkes committed:
Document default TargetOptions settings

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Document TargetOptions changes

- Documented the rationale for implementing the TargetOptions
  class, its methodology, and the changes made.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

nupuruttarwar commented on stratum/docs/target_options.md:
```diff
@@ -0,0 +1,85 @@
+  in common code. The DPDK_TARGET conditionals go against this convention.
+
+- Compile-time conditionals complicate the Bazel build and lead to
+  recompilation when we change change targets. This is undesirable,
```
we change targets?

Right now it's change change

ffoulkes replied:
Oops. Fixed.

nupuruttarwar commented on stratum/hal/bin/tdi/dpdk/dpdk_main.cc:
```diff
@@ -139,11 +140,14 @@ ::util::Status Main(absl::Notification* ready_sync,
 
+  TargetOptions target_options;
+  target_options.allowPipelineOverwrite = false;
+  target_options.pushUpdatedChassisConfig = false;
```
Most likely, we will have to set similar values for ES2K target. I will check with SDE team once and get back to you. This can be taken as a separate PR

nupuruttarwar reviewed (approved):
Minor comment.

LGTM

ffoulkes committed:
Correct error in target_options.md

Signed-off-by: Derek G Foster <derek.foster@intel.com>

satish153 commented on stratum/hal/lib/common/p4_service.cc:
```diff
@@ -57,13 +58,21 @@ namespace hal {
+    : P4Service(mode, switch_interface, auth_policy_checker, error_buffer,
+                TargetOptions::default_target_options) {}
+
+P4Service::P4Service(OperationMode mode, SwitchInterface* switch_interface,
```
Why not have just one constructor like below?
P4Service::P4Service(OperationMode mode, SwitchInterface* switch_interface,
                     AuthPolicyChecker* auth_policy_checker,
                     ErrorBuffer* error_buffer,
                     const TargetOptions& target_options =  TargetOptions::default_target_options)

satish153 commented on stratum/hal/lib/tdi/dpdk/dpdk_hal.h:
```diff
@@ -81,7 +82,8 @@ class DpdkHal final {
                                   absl::Notification* ready_sync = nullptr,
-                                  absl::Notification* done_sync = nullptr)
+                                  absl::Notification* done_sync = nullptr,
+                                  const TargetOptions* target_options = nullptr)
```
const TargetOptions* target_options =TargetOptions::default_target_options?
Since TargetOptions::default_target_options is created to be the default?

ffoulkes replied:
This approach works at the expense of exposing the default target options to sections of the code that don't need to know the implementation details. I prefer passing a null pointer and supplying the defaults at the end of the line.

It also introduces tight coupling in the interface, which could cause problems in the unit tests.

I think I have an alternative approach, but I'd prefer to merge this commit before doing any additional tinkering. I believe it meets Google's "senior" code review principle:

> In general, reviewers should favor approving a CL once it is in a state where it definitely improves the overall code health of the system being worked on, even if the CL isn’t perfect.

satish153 commented on stratum/hal/lib/tdi/es2k/es2k_hal.h:
```diff
@@ -81,7 +82,8 @@ class Es2kHal final {
                                   absl::Notification* ready_sync = nullptr,
-                                  absl::Notification* done_sync = nullptr)
+                                  absl::Notification* done_sync = nullptr,
+                                  const TargetOptions* target_options = nullptr)
```
const TargetOptions* target_options =TargetOptions::default_target_options?
Since TargetOptions::default_target_options is created to be the default?

satish153 reviewed (approved):
Comments are mostly minor. Approved.
